nl: implement -d/--section-delimiter

cakebaker opened the pull request:
This PR implements `-d`/`--section-delimiter` that allows you to define a section delimiter other than the default `\:`. If the specified delimiter is a single character, `:` is added to the delimiter.

tertsdiepraam commented on src/uu/nl/src/nl.rs:
```diff
@@ -106,6 +106,42 @@ impl NumberFormat {
+            return SectionDelimiter::None;
+        }
+
+        match s {
+            _ if s == self.delimiter.repeat(3) => SectionDelimiter::Header,
+            _ if s == self.delimiter.repeat(2) => SectionDelimiter::Body,
+            _ if s == self.delimiter => SectionDelimiter::Footer,
+            _ => SectionDelimiter::None,
+        }
```
Alright, so this is more of a brainstorm than actual feedback, because this expression is very cool and clever but also a bit weird. Maybe it could be:
```rs
if s.trim_left(self.delimiter).is_empty() {
    match s.len() / self.delimiter.len() {
        3 => SectionDelimiter::Header,
        2 => SectionDelimiter::Body,
        1 => SectionDelimiter::Footer,
        _ => SectionDelimiter::None,
}
```
This avoids the allocations of the strings from `repeat` as well.

tertsdiepraam commented on src/uu/nl/src/nl.rs:
```diff
@@ -106,6 +106,42 @@ impl NumberFormat {
+    }
+
+    fn parse(&self, s: &str) -> SectionDelimiter {
+        if self.delimiter.is_empty() {
+            return SectionDelimiter::None;
+        }
```
Does making `self.delimiter` an `Option<SectionDelimiter>` make sense? Or maybe the whole `SectionDelimiterParser`?

tertsdiepraam commented on src/uu/nl/src/nl.rs:
```diff
@@ -106,6 +106,42 @@ impl NumberFormat {
+
+impl SectionDelimiterParser {
+    fn new(delimiter: &str) -> Self {
+        let delimiter = if delimiter.chars().count() == 1 {
```
```suggestion
        let delimiter = if delimiter.len() == 1 {
```

cakebaker replied:
Hm, `len` returns the number of bytes, not the number of chars (for example, `"ä".len()` returns `2`). And my intention was to get the number of chars. On the other hand, using `len` leads to the same output as GNU nl when specifying `ä` as delimiter. Which is incorrect, in my opinion ;-) I guess I have to ask in the GNU coreutils project what the expected output should be in such a case.

tertsdiepraam replied:
Ohh I see. Maybe it should even be unicode width?

cakebaker commented:
@tertsdiepraam I addressed your feedback in the latest push in the following way:

- I replaced `chars().count()` with `len()` and added an additional test case `test_non_ascii_one_char_section_delimiter`
- I rewrote the "weird" match (plus moved the entire function to `SectionDelimiter`)
- the `parse` function now returns an `Option<SectionDelimiter>`

In addition I got rid of the `SectionDelimiterParser`.

tertsdiepraam commented on src/uu/nl/src/nl.rs:
```diff
@@ -121,6 +120,32 @@ impl NumberFormat {
+    Footer,
+}
+
+impl SectionDelimiter {
```
This is beautiful now!

tertsdiepraam commented:
CI failures seem unrelated, but I'm rerunning them just to be sure.

cakebaker commented:
Somehow I overlooked that this PR got approved a while ago :|
